core/src/main/scala/org/apache/spark/PreemptionPolicy.scala

```diff
@@ -0,0 +1,121 @@
+    getCtor(conf)
+      .newInstance(executorIds, removeTimes, preemptedExecutors)
+      .asInstanceOf[PreemptionPolicy]
+  }
+}
+
+/**
+ * The preemptExecutors method is synchronized in the ExecutorAllocationManager (EAM),
+ * so it's not necessary here. But still...beware.
+ *
+ * @param executorIds mutable collection of executorIds (not mutated by default)
+ * @param removeTimes EAM's idle executor -> expiration map (not mutated by default)
+ * @param preemptedExecutors mutable collection of executors to be preempted on the next
+ *                           EAM update. preemptExecutors adds to this set, but EAM acts
+ *                           on and clears the set
+ */
+abstract class PreemptionPolicy(
+    executorIds: mutable.HashSet[String],
+    removeTimes: mutable.HashMap[String, Long],
+    preemptedExecutors: mutable.HashSet[String]) {
+  def preemptExecutors(pe: PreemptExecutors): Unit
+}
+
+/**
+ * DefaultPreemptionPolicy is the default preemption policy selected when
+ * DYN_ALLOCATION_PREEMPTION_POLICY is not set. This policy was created with YARN's
+ * PreemptionMessage symantics in mind. Instead of simply killing YARN's requested
+ * containers, preemptively remove idle executors without cached data. If YARN requests
+ * to preempt more executors than are idle, include the `askedToLeave` set and randomly
+ * select executors to fill out the total preemption count.
+ */
+class DefaultPreemptionPolicy(
+    executorIds: mutable.HashSet[String],
+    removeTimes: mutable.HashMap[String, Long],
+    preemptedExecutors: mutable.HashSet[String]) extends
+  PreemptionPolicy(executorIds, removeTimes, preemptedExecutors) {
+
+  protected[spark] def orderByTimeout(
+      execMap: mutable.HashMap[String, Long],
+      numExecs: Int): Seq[String] = {
+    execMap.toSeq.sortBy(_._2).take(numExecs).map(_._1)
+  }
+
+  /**
+   * find executors to offer up for preemption
+   * favor most idle executors without cached data
+   * NOTE: does not consider non-idle executors
+   */
+  protected[spark] def findPreemptableExecutors(numExecsToKill: Int): Seq[String] = {
+    val blockMaster = SparkEnv.get.blockManager.master
+    val (execsWithCachedData, execsWithoutCachedData) = removeTimes
+      .partition { case (id, _) => blockMaster.hasCachedBlocks(id) }
+    if (execsWithoutCachedData.size < numExecsToKill) {
+      val execsLeftToKill = numExecsToKill - execsWithoutCachedData.size
+      execsWithoutCachedData.keys.toSeq ++
+        orderByTimeout(execsWithCachedData, execsLeftToKill)
+    } else orderByTimeout(execsWithoutCachedData, numExecsToKill)
+  }
+
+  override def preemptExecutors(pe: PreemptExecutors): Unit = {
+    val numExecsToPreempt = pe.forcedToLeave.size + pe.numRequestedContainers
+    val execsToPreempt = mutable.LinkedHashSet(pe.forcedToLeave.toSeq: _*)
+    if (pe.numRequestedContainers > 0) {
+      execsToPreempt ++= findPreemptableExecutors(pe.numRequestedContainers)
```

Comment:
Please document (with a code comment) that YARN will accept alternate sacrificial executors in place of the executors that YARN asked you to preempt; and that what this code is doing is choosing a replacement sacrifice.